File: src/test/java/nl/kabisa/integration/AbstractIT.java
package nl.kabisa.integration;

import org.junit.BeforeClass;
import org.junit.runner.RunWith;
import org.springframework.boot.context.embedded.LocalServerPort;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.*;
import org.springframework.test.context.junit4.SpringRunner;

@RunWith(SpringRunner.class)
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
public class AbstractIT {
    @LocalServerPort
    protected int serverPort;

    @BeforeClass
    public static void setup() {
        System.setProperty("rate.provider1.api.host", "http://localhost:1234");
        System.setProperty("rate.provider2.api.host", "http://localhost:1234");
    }

    protected String apiUrl(String path) {
        return "http://localhost:" + serverPort + path;
    }

    protected HttpEntity<String> jsonEntity(String body) {
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);

        return new HttpEntity<>(body, headers);
    }
}

File: src/test/kotlin/nl/kabisa/integration/AbstractIT.kt
package nl.kabisa.integration

import org.junit.BeforeClass
import org.junit.runner.RunWith
import org.springframework.boot.context.embedded.LocalServerPort
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.test.context.junit4.SpringRunner

@RunWith(SpringRunner::class)
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
open class AbstractIT {
    @LocalServerPort
    protected var serverPort: Int = 0

    protected fun apiUrl(path: String): String {
        return "http://localhost:" + serverPort + path
    }

    protected fun jsonEntity(body: String): HttpEntity<String> {
        val headers = HttpHeaders()
        headers.contentType = MediaType.APPLICATION_JSON

        return HttpEntity(body, headers)
    }

    companion object {

        @BeforeClass @JvmStatic
        fun setup() {
            System.setProperty("rate.provider1.api.host", "http://localhost:1234")
            System.setProperty("rate.provider2.api.host", "http://localhost:1234")
        }
    }
}
